source/logrepl/copyto.go

```diff
@@ -0,0 +1,178 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logrepl
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/conduitio/conduit-connector-postgres/pgutil"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgproto3/v2"
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+)
+
+type CopyDataWriter struct {
+	config            Config
+	fieldDescriptions []pgproto3.FieldDescription
+	decoders          []decoderValue
+	connInfo          *pgtype.ConnInfo
+	messages          chan sdk.Record
+	done              chan struct{}
+}
+
+// NewCopyDataWriter builds a new CopyDataWriter from a postgres connection.
+func NewCopyDataWriter(ctx context.Context, conn *pgx.Conn, config Config) (*CopyDataWriter, error) {
+	fds, err := getFieldDescriptions(ctx, conn.PgConn(), []string{config.TableName})
+	if err != nil {
+		return nil, err
+	}
+
+	decoders := make([]decoderValue, len(fds))
+	for i, fd := range fds {
+		decoders[i] = oidToDecoderValue(pgtype.OID(fd.DataTypeOID))
+	}
+
+	c := &CopyDataWriter{
+		config:            config,
+		fieldDescriptions: fds,
+		decoders:          decoders,
+		connInfo:          conn.ConnInfo(),
+		messages:          make(chan sdk.Record, 1),
+		done:              make(chan struct{}, 1),
+	}
+
+	return c, nil
+}
+
+// Copy is meant to be called concurrently after a writer is created.
+func (c *CopyDataWriter) Copy(ctx context.Context, conn *pgx.Conn) {
+	copyquery := fmt.Sprintf("COPY %s TO STDOUT", c.config.TableName)
+	_, err := conn.PgConn().CopyTo(ctx, c, copyquery)
+	if err != nil {
+		fmt.Printf("failed to copy data from table: %v", err)
+	}
+	c.done <- struct{}{}
+}
+
+func (c *CopyDataWriter) Write(line []byte) (n int, err error) {
+	tokens := bytes.Split(line, []byte{'\t'})
```

Comment:
**Most important comment**

We need to make sure that we parse the line correctly, specifically see the documentation for [Copy - Text Format](https://www.postgresql.org/docs/current/sql-copy.html). We currently only handle the null string `\N`, but there are others (e.g. `\\`, `\b`, `\f`, `\n`, `\r`, `\t`, `\v`, `\digits`, `\xdigits`). That's also the reason why `bytea` values are not decoded correctly (the `bytea` text decoder will work once we replace `\\` with `\`). It's not as easy as just replacing values though, for example `\\n` is actually the string `\n` and not a new line. It would be almost necessary to have tests for these values to know we are handling them correctly.